signac/contrib/migration/__init__.py

```diff
@@ -5,102 +5,135 @@
+            # not terribly consequential.
+            config = _CONFIG_LOADERS[guess.public](root_directory)
             break
+        except Exception:
+            # The load failed, go to the next
+            pass
     else:
-        raise RuntimeError("Unable to determine project configuration file.")
-    config.update(kwargs)
-    config.write()
-    _reload_project_config(project)
-
-
-@contextmanager
-def _lock_for_migration(project):
-    lock = FileLock(project.fn(FN_MIGRATION_LOCKFILE))
+        raise RuntimeError("Unable to load config file.")
     try:
-        with lock:
-            yield
-    finally:
-        try:
-            os.unlink(lock.lock_file)
-        except FileNotFoundError:
-            pass
+        return version.parse(config["schema_version"])
+    except KeyError:
+        # The default schema version is version 0.
+        return version.parse("0")
 
 
-def _collect_migrations(project):
+def _collect_migrations(root_directory):
     schema_version = version.parse(SCHEMA_VERSION)
 
-    def get_config_schema_version():
-        return version.parse(project._config["schema_version"])
-
-    if get_config_schema_version() > schema_version:
+    current_schema_version = _get_config_schema_version(root_directory, SCHEMA_VERSION)
+    if current_schema_version > schema_version:
         # Project config schema version is newer and therefore not supported.
         raise RuntimeError(
             "The signac schema version used by this project is {}, but signac {} "
             "only supports up to schema version {}. Try updating signac.".format(
-                get_config_schema_version(), __version__, SCHEMA_VERSION
+                current_schema_version, __version__, SCHEMA_VERSION
             )
         )
 
-    while get_config_schema_version() < schema_version:
-        for (origin, destination), migration in MIGRATIONS.items():
-            if version.parse(origin) == get_config_schema_version():
+    guess = current_schema_version
+    while _get_config_schema_version(root_directory, guess) < schema_version:
+        for (origin, destination), migration in _MIGRATIONS.items():
+            if version.parse(origin) == _get_config_schema_version(
+                root_directory, guess
+            ):
                 yield (origin, destination), migration
+                guess = destination
                 break
         else:
             raise RuntimeError(
                 "The signac schema version used by this project is {}, but signac {} "
                 "uses schema version {} and does not know how to migrate.".format(
-                    get_config_schema_version(), __version__, schema_version
+                    _get_config_schema_version(root_directory, guess),
+                    __version__,
+                    schema_version,
                 )
             )
 
 
-def apply_migrations(project):
+def apply_migrations(root_directory):
     """Apply migrations to a project."""
```

Comment:
Is this method supposed to be public or private? If public, it needs more docstrings (might not hurt to expand the docstring even if it's private). If private, it needs to be renamed to `_apply_migrations`.

Reply:
It is public. This is the public Python API for migrating, the analog to the CLI `signac migrate` (which is currently disabled but should be exposed soon). I can more documentation.